File: pset6/sentimental-readability/readability.py
from cs50 import get_string


def main():
    # Get text
    text = get_string("Text: ")

    # Calculate letters, words, and sentences
    letters, words, sentences = count_text_data(text)

    # Calculate index
    index = calculate_index(letters, words, sentences)

    # Print index
    print_grade(index)


# Calculate letters, words, and sentences


def count_text_data(text):
    letters = 0
    words = 1
    sentences = 0

    for c in text:
        if c == " ":
            words += 1
        elif c in ['.', '?', '!']:
            sentences += 1
        elif c.isalpha():
            letters += 1

    return [letters, words, sentences]


# Calculate index


def calculate_index(letters, words, sentences):

    L = letters / words * 100
    S = sentences / words * 100

    return 0.0588 * L - 0.296 * S - 15.8

# Print grade


def print_grade(index):
    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print("Grade ", round(index))


main()

File: pset2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <math.h>

float calculate_grade(string text);

int main(void)
{
    string text = get_string("Text: ");

    float grade = calculate_grade(text);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        int rounded_grade = round(grade);
        printf("Grade %i\n", rounded_grade);
    }
}

float calculate_grade(string text)
{
    // Initialize letters words and sentences
    int letters = 0;
    int words = 0;
    int sentences = 0;

    int i = 0;
    while (text[i] != '\0')
    {
        char c = text[i];

        // If it's a letter, add to letters
        if (isalpha(c))
        {
            letters++;
        }

        // If it's a space, add to words
        if (c == ' ')
        {
            words++;
        }

        // If its a . ! or ?, add to sentences.
        if (c == '.' || c == '!' || c == '?')
        {
            sentences++;
        }
        i++;
    }

    // We add one more word per the logic
    words++;

    // Calculate formula's variables
    float letter_avg = (float) letters / words * 100;
    float sentence_avg = (float) sentences / words * 100;

    float grade = 0.0588 * letter_avg - 0.296 * sentence_avg - 15.8;

    return grade;
}